### PR TITLE
feat: Propagate exceptions to AfterToolCallEvent for decorated tools (#1565)

### PR DESCRIPTION
**Pull request by charles-dyfis-net**
### Motivation

Hook providers need the ability to distinguish between expected errors (like validation failures that the model should adjust for and retry) and unexpected errors (like assertion failures or configuration errors that should fail the task immediately). Currently, `AfterToolCallEvent.exception` is always `None` for `@tool`-decorated functions because the decorator catches all exceptions and converts them to error results before the executor can see them. This forces all tool errors to be returned to the model, even when propagating the exception would be more appropriate.

Resolves: #1565

### Public API Changes

`ToolResultEvent` now accepts an optional `exception` parameter, and `AfterToolCallEvent.exception` is populated for decorated tools:

```python
# Before: exception is always None for decorated tools
class MyHook(HookProvider):
    def register_hooks(self, registry, **kwargs):
        registry.add_callback(AfterToolCallEvent, self._check)

    def _check(self, event: AfterToolCallEvent):
        # event.exception is always None for @tool functions
        # even when the tool raised an exception
        pass

# After: exception is available for inspection
class PropagateUnexpectedExceptions(HookProvider):
    def __init__(self, allowed_exceptions=(ValueError,)):
        self.allowed_exceptions = allowed_exceptions

    def register_hooks(self, registry, **kwargs):
        registry.add_callback(AfterToolCallEvent, self._check)

    def _check(self, event: AfterToolCallEvent):
        if event.exception is None:
            return
        if isinstance(event.exception, self.allowed_exceptions):
            return  # Let model retry
        raise event.exception  # Propagate unexpected errors
```

The error result is still returned to the model by default; hooks must explicitly re-raise if they want to propagate.

### Use Cases

- **Fail-fast on unexpected errors**: Propagate `AssertionError`, `ConfigurationError`, etc. instead of letting the model retry futilely
- **Error classification**: Log or handle different exception types differently based on their class
- **Selective retry**: Allow `ValueError` to return to the model while propagating other exceptions

## Related Issues

#1565

## Documentation PR

strands-agents/docs#482

## Type of Change

New feature

## Testing

How have you tested the change?  Verify that the changes do not break functionality or introduce warnings in consuming repositories: agents-docs, agents-tools, agents-cli

- [X] I ran `hatch run prepare`

## Checklist
- [ ] I have read the CONTRIBUTING document
- [X] I have added any necessary tests that prove my fix is effective or my feature works
- [X] I have updated the documentation accordingly
- [X] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/strands/tools/executors/test_executor.py [hunk 1]
```diff
@@ -483,6 +483,74 @@ async def test_executor_stream_updates_invocation_state_with_agent(
...
+    tool_use = {"name": "decorated_error_tool", "toolUseId": "1", "input": {}}
+
+    stream = executor._stream(agent, tool_use, tool_results, invocation_state)
+    await alist(stream)
+
+    after_event = hook_events[-1]
+    assert isinstance(after_event, AfterToolCallEvent)
+    assert after_event.exception is not None
+    assert isinstance(after_event.exception, ValueError)
+    assert "decorated tool error" in str(after_event.exception)
```
Comment (zastrowm) [inline tests/strands/tools/executors/test_executor.py]: Can this be simplified to 
```suggestion
    assert after_event.exception == ValueError("decorated tool error")
```

Or something similar?  We should be able to verify the exact exception rather than "decorated tool error" in str(after_event.exception)", given that we control it.

Alternatively, do:

```
exception = ValueError("decorated tool error")
```

early on, and then check that it's:

```
assert after_event.exception == exception
```

File: tests/strands/tools/executors/test_executor.py [hunk 2]
```diff
@@ -483,6 +483,74 @@ async def test_executor_stream_updates_invocation_state_with_agent(
...
+    tool_use = {"name": "runtime_error_tool", "toolUseId": "1", "input": {}}
+
+    stream = executor._stream(agent, tool_use, tool_results, invocation_state)
+    await alist(stream)
+
+    after_event = hook_events[-1]
+    assert isinstance(after_event, AfterToolCallEvent)
+    assert after_event.exception is not None
+    assert isinstance(after_event.exception, RuntimeError)
+    assert "runtime error from decorated tool" in str(after_event.exception)
```
Comment (zastrowm) [inline tests/strands/tools/executors/test_executor.py]: Along the lines of my earlier comment, let's simplify this assertion

File: tests/strands/tools/executors/test_executor.py [hunk 3]
```diff
@@ -483,6 +483,74 @@ async def test_executor_stream_updates_invocation_state_with_agent(
...
+    agent.tool_registry.register_tool(success_tool)
+    tool_use = {"name": "success_decorated_tool", "toolUseId": "1", "input": {}}
+
+    stream = executor._stream(agent, tool_use, tool_results, invocation_state)
+    await alist(stream)
+
+    after_event = hook_events[-1]
+    assert isinstance(after_event, AfterToolCallEvent)
+    assert after_event.exception is None
+    assert after_event.result["status"] == "success"
```
Comment (zastrowm) [inline tests/strands/tools/executors/test_executor.py:551]: nit: newline after this I think

File: tests/strands/tools/executors/test_executor.py [hunk 4]
```diff
@@ -483,6 +483,74 @@ async def test_executor_stream_updates_invocation_state_with_agent(
 
 
 @pytest.mark.asyncio
+async def test_executor_stream_decorated_tool_exception_in_hook(
```
Comment (zastrowm) [inline tests/strands/tools/executors/test_executor.py:486]: Can we add a test case that ensures that exception is None, but the result is failed if the decorator returns an error state without throwing?  I'd like to ensure that behavior is solidified in a test

**Comment by zastrowm** [comment]
Looks like there are also some linting failures;  `uv run hatch run prepare` is my goto for fixing/checking

File: src/strands/tools/executors/_executor.py [hunk 5]
```diff
@@ -239,7 +244,7 @@ async def _stream(
                 result = cast(ToolResult, event)
 
                 after_event, _ = await ToolExecutor._invoke_after_tool_call_hook(
-                    agent, selected_tool, tool_use, invocation_state, result
+                    agent, selected_tool, tool_use, invocation_state, result, exception=exception
```
Comment (zastrowm) [inline src/strands/tools/executors/_executor.py:247]: Sorry I missed this before; can this just be
```suggestion
                    agent, selected_tool, tool_use, invocation_state, result, exception=event.exception
```

Removing the need to declare exception above?

**Review by zastrowm — APPROVED** [review]

File: tests/strands/tools/test_decorator.py [hunk 6]
```diff
@@ -1823,3 +1823,80 @@ def test_tool_decorator_annotated_field_with_inner_default():
...
+
+
+@pytest.mark.asyncio
+async def test_tool_result_event_carries_exception_runtime_error(alist):
+    """Test that ToolResultEvent carries exception when tool raises RuntimeError."""
+
+    @strands.tool
+    def error_tool():
+        """Tool that raises a RuntimeError."""
+        raise RuntimeError("test runtime error")
```
Comment (zastrowm) [inline tests/strands/tools/test_decorator.py:1835]: Can we use the same technique here where we put this in a variable and assert on the exact value later

(and below too)

File: src/strands/tools/executors/_executor.py (hunk 5 above)
Comment (dbschmigelski) [reply src/strands/tools/executors/_executor.py:247]: I think the issue is that we currently reassign the value of event 

```
                        # below the last "event" must point to the tool_result
                        event = event.tool_result
```

File: src/strands/types/_events.py [hunk 7]
```diff
@@ -274,15 +274,29 @@ def prepare(self, invocation_state: dict) -> None:
 
 
 class ToolResultEvent(TypedEvent):
-    """Event emitted when a tool execution completes."""
+    """Event emitted when a tool execution completes.
 
-    def __init__(self, tool_result: ToolResult) -> None:
-        """Initialize with the completed tool result.
+    Stores the full Exception object as an instance attribute for debugging while
```
Comment (mkmeral) [inline src/strands/types/_events.py]: nit: we do not want implementation details in our docstrings

**Review by pgrayy — APPROVED** [review]

File: src/strands/types/_events.py (hunk 7 above)
Comment (dbschmigelski) [reply src/strands/types/_events.py]: removed

**Comment by charles-dyfis-net** [comment]
Any further attention needed from my end here?

**Comment by pgrayy** [comment]
> Any further attention needed from my end here?

Apologies, nothing else needed. Went ahead and merged.
